### PR TITLE
ED-25519 Derived Extended Private Keys

### LINKED ISSUE #1320: Add SLIP 23 To Key Management
## Description

Add a SLIP 23 implementation to the `common/keymanagement` project.

A basic SLIP-23 implementation exists in the [Topl/Crypto Repository](https://github.com/Topl/Crypto/blob/master/src/main/scala/crypto/Slip23.scala).

## Acceptance Criteria

- SLIP 23 implementation is added in `common/keymanagement`
- SLIP 23 unit tests are added and are passing

## Rework

- [ ] add comments to the ByteVector class for more explanation
- [ ] add test vectors from rust-cardano `hdwallet.rs` implementation
- [ ] re-investiage public key derivation from an extended private key (see #1426)
- [ ] change mnemonic from accepting an `Option[String]` to a `String`

### LINKED ISSUE #1426: Add Public Key Derivation From Extended Public Key
## Description

The current BIP32-ED25519 implementation located in key management allows deriving a child private key from a parent private key as outlined in the BIP32-ED25519 specification. However, it does not implement the public key derivation from a public key which the spec also allows for.

## Acceptance Criteria

Valid ED-25519 soft-index public keys can be generated from a parent public key.

Helpful Hints and resources located at [BIP32-ED25519 Public Key -> Public Key Derivation](https://topl.atlassian.net/l/c/sMEVgfd7).

### LINKED ISSUE #1477: Add BIP32-ED25519 Test Vectors From rust-cardano HD Wallet Project
## Description

The [rust-cardano](https://github.com/input-output-hk/rust-cardano) project has easy to understand test vectors for BIP32-ED25519 located in the `cardano/src/hdwallet.rs` file. These should be replicated in the `ExtendedPrivateKeyEd25519Spec` in the `attestation` package.

## Acceptance Criteria

- tests from `rust-cardano` are run as unit tests in Bifrost

### LINKED ISSUE #1526: SLIP-0023 Cardano Icarus master node derivation
## Description

Implement [SLIP-0023 Cardano Icarus master node derivation](https://github.com/satoshilabs/slips/blob/master/slip-0023.md) so that key pairs will be compatible with all wallets that currently accept cardano.

This scheme requires changes to existing work on BIP-0039 mnemonics and SLIP-0023 Cardano universal master node derivation.

No changes will be necessary to existing work on child key derivation.

## Acceptance Criteria

- an extended private key can be generated from a BIP-0039 mnemonic that follows the SLIP-0023 Cardano Icarus master node derivation spec
- test vectors exist for the entire HD-wallet stack and are documented

### PR DESCRIPTION
**Pull request by b-vennes**
## Purpose

Most blockchains have a method for creating hundreds of thousands of signing key pairs and addresses from a root seed value (usually generated by a BIP-39 mnemonic or similar). Bifrost/Brambl needs a standard derived key implementation that lets users input some entropy (a mnemonic phrase) and get back a deterministic set of keys and addresses. 

## Approach

- Implements SLIP-23/BIP32-ED25519 protocols for deriving private ED-25519 signing keys from a root seed
- implements signing and serialization functions for `ExtendedPrivateKeyEd25519` keys so that they can be used as valid `Secrets` in the future

## Testing

- adds `ExtendedPrivateKeyEd25519Spec` with test vectors from [HD Wallet Protocols and Test Vectors document](https://topl.atlassian.net/wiki/external/294813812/Y2VkZTEzYzg4MzA2NGViYTg4NzExMTRmMWU3MGQ2ZWI?atlOrigin=eyJpIjoiYTU3OGExYWJhMTVjNGRhZWIwNDE2NGY3ZmI4MjI1MzciLCJwIjoiYyJ9)

## Tickets

- closes #1320
- closes #1426
- closes #1477
- closes #1526

### CODE REVIEW

File: common/src/main/scala/co/topl/attestation/keyManagement/derivedKeys/ExtendedPrivateKeyEd25519.scala [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+    PublicKey(pk)
+  }
+
+  /**
+   * Deterministically derives a child `ExtendedPrivateKey` at the given index from this extended private key.
+   * @param index the index of the child key to derive
+   * @return the derived `ExtendedPrivateKey`
+   */
+  def deriveChildKey(
```
Comment (scasplte2) [inline common/src/main/scala/co/topl/attestation/keyManagement/derivedKeys/ExtendedPrivateKeyEd25519.scala:55]: just my two-cents but if I were looking at this 6 months later I might like to know that this function basically corresponds to section V.C

File: common/src/main/scala/co/topl/attestation/keyManagement/derivedKeys/package.scala [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package co.topl.attestation.keyManagement
+
+import org.bouncycastle.crypto.digests.SHA512Digest
+import org.bouncycastle.crypto.macs.HMac
+import org.bouncycastle.crypto.params.KeyParameter
+import scodec.bits.ByteVector
+
+package object derivedKeys {
+
+  def hmac512WithKey(key: ByteVector, data: ByteVector): ByteVector = {
```
Comment (scasplte2) [inline common/src/main/scala/co/topl/attestation/keyManagement/derivedKeys/package.scala:10]: could you clarify for me the interaction between this function and `object Pbkdf2Sha512`

**Review by scasplte2 — CHANGES_REQUESTED** [review]
- Reproduce and document test vectors from rust-cardano or jormungandr
- reconsider public parent-> public child
- add more comments to SizedByteVector typeclass

File: common/src/main/scala/co/topl/attestation/keyManagement/derivedKeys/ExtendedPrivateKeyEd25519.scala [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+   * @return a root `ExtendedPrivateKey`
+   */
+  def fromSeed(seed: Array[Byte]): ExtendedPrivateKeyEd25519 = {
+    val i =
+      hmac512WithKey(ByteVector.view("ed25519 cardano seed".getBytes(StandardCharsets.UTF_8)), ByteVector.view(seed))
+
+    val iLeft = i.slice(0, 32)
+    val iRight = i.slice(32, 64)
+
+    val k = sha512.hash(iLeft.toArray).value
```
Comment (scasplte2) [inline common/src/main/scala/co/topl/attestation/keyManagement/derivedKeys/ExtendedPrivateKeyEd25519.scala:179]: might note that this bit (particularly the bit flipping) is specified in V.A

File: common/src/main/scala/co/topl/attestation/keyManagement/derivedKeys/ExtendedPrivateKeyEd25519.scala [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    k(0) = (k(0) & 0xf8).toByte
+    k(31) = ((k(31) & 0x1f) | 0x40).toByte
+
+    ExtendedPrivateKeyEd25519(
+      SizedByteVector[ByteVector32].fit(k.slice(0, 32), ByteOrdering.LittleEndian),
+      SizedByteVector[ByteVector32].fit(k.slice(32, 64), ByteOrdering.LittleEndian),
+      SizedByteVector[ByteVector32].fit(iRight, ByteOrdering.LittleEndian),
```
Comment (scasplte2) [inline common/src/main/scala/co/topl/attestation/keyManagement/derivedKeys/ExtendedPrivateKeyEd25519.scala:186]: I'm a little confused by this root chain code. It looks to me that in V.A the root chain code would be c <- H_{256}(0x01||kbar)

File: common/src/main/scala/co/topl/utils/SizedByteCollection.scala [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+import java.nio.{ByteBuffer, ByteOrder}
+import scala.language.implicitConversions
+
+/**
+ * Type-class for working with a collection of bytes with a specific size.
+ * @tparam T the underlying type of the collection
+ */
+@typeclass
+trait SizedByteCollection[T] {
```
Comment (scasplte2) [inline common/src/main/scala/co/topl/utils/SizedByteCollection.scala:16]: 👍

**Review by scasplte2 — DISMISSED** [review]
👍

File: common/src/main/scala/co/topl/attestation/keyManagement/mnemonic/Mnemonic.scala [hunk 6]
```diff
@@ -180,17 +211,40 @@ object Mnemonic {
...
-        validatedPhrase.getBytes("UTF-8"),
-        ("mnemonic" + password.getOrElse("")).getBytes("UTF-8"),
-        64,
-        2048
-      )
+  private def entropyFromValidatedPhrase(
+    words:     IndexedSeq[String],
+    wordsList: IndexedSeq[String],
+    size:      MnemonicSize
+  ): Array[Byte] =
```
Comment (b-vennes) [inline common/src/main/scala/co/topl/attestation/keyManagement/mnemonic/Mnemonic.scala:240]: TODO: Update all the types in Mnemonic to use aliases

File: common/src/main/scala/co/topl/attestation/keyManagement/mnemonic/mnemonic.scala [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  private[mnemonic] val byteLen = 8
+  private[mnemonic] val indexLen = 11
+
+  /**
+   * Converts a mnemonic index into a binary representation.
+   * @param i the index to convert
+   * @return the binary representation as a `String`
+   */
+  private[mnemonic] def toBinaryIndex(i: Int): String =
+    String.format("%11s", BigInt(i).toString(2)).replace(' ', '0')
```
Comment (b-vennes) [inline common/src/main/scala/co/topl/attestation/keyManagement/mnemonic/mnemonic.scala:16]: TODO: note why this is %11s

File: common/src/main/scala/co/topl/utils/SizedByteCollection.scala (hunk 5 above)
Comment (b-vennes) [reply common/src/main/scala/co/topl/utils/SizedByteCollection.scala:16]: TODO Consider naming:
- SizedBytes
- SizedByteVector
- something short

File: common/src/main/scala/co/topl/attestation/keyManagement/derivedKeys/ExtendedPrivateKeyEd25519.scala [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+
+  /**
+   * Deterministically derives a child `ExtendedPrivateKey` at the given index from this extended private key.
+   *
+   * Follows section V.C from BIP32-ED25519 spec.
+   *
+   * @param index the index of the child key to derive
+   * @return the derived `ExtendedPrivateKey`
+   */
+  def derive(
```
Comment (scasplte2) [inline common/src/main/scala/co/topl/attestation/keyManagement/derivedKeys/ExtendedPrivateKeyEd25519.scala:59]: @SeanCheatham this is a good example of what the key manager (keyRing, key chain, or whatever) will need to manage. Some entity will need to hold onto the top level key that is used to generate sub keys for the levels of the tree that we'll talk about in journal club today.

**Comment by scasplte2** [comment]
@b-vennes very nicely done! The change to Entropy and Phrase makes it much cleaner to me. I'm happy with this as is now so I am ready to merge.

**Review by scasplte2 — APPROVED** [review]
looked over the changes and nothing else to add 👍

**Pull request merged**